File: velocity-sdk-go/advanced.go
package velocity

import (
	"context"
	"fmt"
)

// ─── Workflow Update ────────────────────────────────────────────────────────────

// UpdateOptions contains options for updating a workflow.
type UpdateOptions struct {
	UpdateName string
	Args       interface{}
	WaitPolicy string // "ACCEPTED", "COMPLETED"
}

// UpdateResult represents the result of a workflow update.
type UpdateResult struct {
	UpdateID string
	Status   string // "ACCEPTED", "COMPLETED", "REJECTED"
	Result   interface{}
}

// Update sends an update to a running workflow.
func (c *Client) Update(ctx context.Context, workflowID string, options UpdateOptions) (*UpdateResult, error) {
	return c.conn.UpdateWorkflow(ctx, &UpdateWorkflowRequest{
		Namespace:  c.namespace,
		WorkflowID: workflowID,
		UpdateName: options.UpdateName,
		Args:       options.Args,
		WaitPolicy: options.WaitPolicy,
	})
}

// Update sends an update to the workflow via handle.
func (h *WorkflowHandle) Update(ctx context.Context, updateName string, args ...interface{}) (*UpdateResult, error) {
	return h.client.Update(ctx, h.workflowID, UpdateOptions{
		UpdateName: updateName,
		Args:       args,
	})
}

// ─── Workflow Reset ─────────────────────────────────────────────────────────────

// ResetOptions contains options for resetting a workflow.
type ResetOptions struct {
	ResetEventID int64
	Reason       string
}

// Reset resets a workflow to a specific event ID.
func (c *Client) Reset(ctx context.Context, workflowID string, options ResetOptions) (string, error) {
	return c.conn.ResetWorkflow(ctx, &ResetWorkflowRequest{
		Namespace:  c.namespace,
		WorkflowID: workflowID,
		ResetEventID: options.ResetEventID,
		Reason:     options.Reason,
	})
}

// Reset resets the workflow via handle.
func (h *WorkflowHandle) Reset(ctx context.Context, resetEventID int64, reason string) (string, error) {
	return h.client.Reset(ctx, h.workflowID, ResetOptions{
		ResetEventID: resetEventID,
		Reason:       reason,
	})
}

// ─── Schedule Client ────────────────────────────────────────────────────────────

// ScheduleOptions contains options for creating a schedule.
type ScheduleOptions struct {
	ScheduleID   string
	WorkflowType string
	TaskQueue    string
	CronSchedule string
	Input        interface{}
	Enabled      bool
}

// ScheduleClient provides schedule management operations.
type ScheduleClient struct {
	client *Client
}

// NewScheduleClient creates a new ScheduleClient from an existing Client.
func NewScheduleClient(client *Client) *ScheduleClient {
	return &ScheduleClient{client: client}
}

// Create creates a new schedule.
func (sc *ScheduleClient) Create(ctx context.Context, options ScheduleOptions) (string, error) {
	return sc.client.conn.CreateSchedule(ctx, &CreateScheduleRequest{
		Namespace:    sc.client.namespace,
		ScheduleID:   options.ScheduleID,
		WorkflowType: options.WorkflowType,
		TaskQueue:    options.TaskQueue,
		CronSchedule: options.CronSchedule,
		Input:        options.Input,
		Enabled:      options.Enabled,
	})
}

// Describe returns schedule details.
func (sc *ScheduleClient) Describe(ctx context.Context, scheduleID string) (*Schedule, error) {
	return sc.client.conn.DescribeSchedule(ctx, &DescribeScheduleRequest{
		Namespace:  sc.client.namespace,
		ScheduleID: scheduleID,
	})
}

// List lists all schedules.
func (sc *ScheduleClient) List(ctx context.Context) ([]*Schedule, error) {
	return sc.client.conn.ListSchedules(ctx, &ListSchedulesRequest{
		Namespace: sc.client.namespace,
	})
}

// Update updates a schedule.
func (sc *ScheduleClient) Update(ctx context.Context, scheduleID string, options ScheduleOptions) error {
	return sc.client.conn.UpdateSchedule(ctx, &UpdateScheduleRequest{
		Namespace:    sc.client.namespace,
		ScheduleID:   scheduleID,
		WorkflowType: options.WorkflowType,
		CronSchedule: options.CronSchedule,
		Enabled:      options.Enabled,
	})
}

// Delete deletes a schedule.
func (sc *ScheduleClient) Delete(ctx context.Context, scheduleID string) error {
	return sc.client.conn.DeleteSchedule(ctx, &DeleteScheduleRequest{
		Namespace:  sc.client.namespace,
		ScheduleID: scheduleID,
	})
}

// Pause pauses a schedule.
func (sc *ScheduleClient) Pause(ctx context.Context, scheduleID string) error {
	return sc.client.conn.UpdateSchedule(ctx, &UpdateScheduleRequest{
		Namespace:  sc.client.namespace,
		ScheduleID: scheduleID,
		Enabled:    false,
	})
}

// Unpause resumes a schedule.
func (sc *ScheduleClient) Unpause(ctx context.Context, scheduleID string) error {
	return sc.client.conn.UpdateSchedule(ctx, &UpdateScheduleRequest{
		Namespace:  sc.client.namespace,
		ScheduleID: scheduleID,
		Enabled:    true,
	})
}

// ─── Search Attributes ──────────────────────────────────────────────────────────

// SearchAttributesClient provides search attribute operations.
type SearchAttributesClient struct {
	client *Client
}

// NewSearchAttributesClient creates a new SearchAttributesClient.
func NewSearchAttributesClient(client *Client) *SearchAttributesClient {
	return &SearchAttributesClient{client: client}
}

// UpsertSearchAttributes upserts search attributes for a workflow execution.
func (sac *SearchAttributesClient) Upsert(ctx context.Context, workflowID string, attributes map[string]interface{}) error {
	return sac.client.conn.UpsertSearchAttributes(ctx, &UpsertSearchAttributesRequest{
		Namespace:  sac.client.namespace,
		WorkflowID: workflowID,
		Attributes: attributes,
	})
}

// ListWorkflowExecutions lists workflows matching a search query.
func (sac *SearchAttributesClient) ListWorkflows(ctx context.Context, query string) ([]*WorkflowExecution, error) {
	return sac.client.conn.ListWorkflowExecutions(ctx, &ListWorkflowExecutionsRequest{
		Namespace: sac.client.namespace,
		Query:     query,
	})
}

// CountWorkflowExecutions counts workflows matching a search query.
func (sac *SearchAttributesClient) CountWorkflows(ctx context.Context, query string) (int64, error) {
	return sac.client.conn.CountWorkflowExecutions(ctx, &CountWorkflowExecutionsRequest{
		Namespace: sac.client.namespace,
		Query:     query,
	})
}

// ─── Continue-as-New ────────────────────────────────────────────────────────────

// ContinueAsNewOptions contains options for continuing a workflow as new.
type ContinueAsNewOptions struct {
	WorkflowType string
	TaskQueue    string
	Input        interface{}
	RunTimeout   int64 // milliseconds
	TaskTimeout  int64 // milliseconds
	RetryPolicy  *RetryPolicy
	Memo         map[string]interface{}
}

// ContinueAsNewError is a special error used to signal the worker to continue-as-new.
type ContinueAsNewError struct {
	Options ContinueAsNewOptions
}

func (e *ContinueAsNewError) Error() string {
	return fmt.Sprintf("continue-as-new: %s", e.Options.WorkflowType)
}

// NewContinueAsNewError creates a ContinueAsNewError that instructs the worker
// to continue the workflow as a new execution. This is the Temporal pattern for
// long-running workflows that need to reset their history.
func NewContinueAsNewError(options ContinueAsNewOptions) *ContinueAsNewError {
	return &ContinueAsNewError{Options: options}
}

// ─── Batch Operations ───────────────────────────────────────────────────────────

// BatchOperationOptions contains options for starting a batch operation.
type BatchOperationOptions struct {
	Operation   string // "terminate", "cancel", "signal", "delete"
	Query       string // Visibility query to select workflows
	SignalName  string // Required if Operation is "signal"
	SignalInput interface{}
	Reason      string
}

// BatchOperationClient provides batch operation management.
type BatchOperationClient struct {
	client *Client
}

// NewBatchOperationClient creates a new BatchOperationClient.
func NewBatchOperationClient(client *Client) *BatchOperationClient {
	return &BatchOperationClient{client: client}
}

// Start starts a batch operation.
func (bc *BatchOperationClient) Start(ctx context.Context, options BatchOperationOptions) (string, error) {
	return bc.client.conn.StartBatchOperation(ctx, &StartBatchOperationRequest{
		Namespace:  bc.client.namespace,
		Operation:  options.Operation,
		Query:      options.Query,
		SignalName: options.SignalName,
		Reason:     options.Reason,
	})
}

// Describe returns batch operation details.
func (bc *BatchOperationClient) Describe(ctx context.Context, jobID string) (*BatchOperation, error) {
	return bc.client.conn.DescribeBatchOperation(ctx, &DescribeBatchOperationRequest{
		Namespace: bc.client.namespace,
		JobID:     jobID,
	})
}

// List lists batch operations.
func (bc *BatchOperationClient) List(ctx context.Context) ([]*BatchOperation, error) {
	return bc.client.conn.ListBatchOperations(ctx, &ListBatchOperationsRequest{
		Namespace: bc.client.namespace,
	})
}

// ─── Saga Orchestration ─────────────────────────────────────────────────────────

// SagaStep represents a single step in a saga with an execution function and
// a compensating function for rollback.
type SagaStep struct {
	Execute     func(ctx context.Context) (interface{}, error)
	Compensate  func(ctx context.Context) error
	Name        string
}

// Saga provides saga orchestration for multi-step workflows with compensating
// transactions. If any step fails, previously completed steps are rolled back
// in reverse order (Temporal Saga pattern).
type Saga struct {
	steps       []SagaStep
	completed   []SagaStep
	results     []interface{}
}

// NewSaga creates a new Saga orchestrator.
func NewSaga() *Saga {
	return &Saga{}
}

// AddStep adds a step to the saga.
func (s *Saga) AddStep(name string, execute func(ctx context.Context) (interface{}, error), compensate func(ctx context.Context) error) {
	s.steps = append(s.steps, SagaStep{
		Execute:    execute,
		Compensate: compensate,
		Name:       name,
	})
}

// Execute runs all saga steps in order. If any step fails, previously completed
// steps are compensated in reverse order. Returns the results of successful steps
// and the error from the failing step (if any).
func (s *Saga) Execute(ctx context.Context) ([]interface{}, error) {
	for _, step := range s.steps {
		result, err := step.Execute(ctx)
		if err != nil {
			// Compensate in reverse order
			s.compensate(ctx)
			return s.results, fmt.Errorf("saga step %q failed: %w", step.Name, err)
		}
		s.completed = append(s.completed, step)
		s.results = append(s.results, result)
	}
	return s.results, nil
}

// compensate runs compensating transactions in reverse order.
func (s *Saga) compensate(ctx context.Context) {
	for i := len(s.completed) - 1; i >= 0; i-- {
		_ = s.completed[i].Compensate(ctx)
	}
}

// ─── Connection Request/Response Types ──────────────────────────────────────────

type UpdateWorkflowRequest struct {
	Namespace  string
	WorkflowID string
	UpdateName string
	Args       interface{}
	WaitPolicy string
}

type ResetWorkflowRequest struct {
	Namespace    string
	WorkflowID   string
	ResetEventID int64
	Reason       string
}

type CreateScheduleRequest struct {
	Namespace    string
	ScheduleID   string
	WorkflowType string
	TaskQueue    string
	CronSchedule string
	Input        interface{}
	Enabled      bool
}

type DescribeScheduleRequest struct {
	Namespace  string
	ScheduleID string
}

type ListSchedulesRequest struct {
	Namespace string
}

type UpdateScheduleRequest struct {
	Namespace    string
	ScheduleID   string
	WorkflowType string
	CronSchedule string
	Enabled      bool
}

type DeleteScheduleRequest struct {
	Namespace  string
	ScheduleID string
}

type UpsertSearchAttributesRequest struct {
	Namespace  string
	WorkflowID string
	Attributes map[string]interface{}
}

type ListWorkflowExecutionsRequest struct {
	Namespace string
	Query     string
}

type CountWorkflowExecutionsRequest struct {
	Namespace string
	Query     string
}

type StartBatchOperationRequest struct {
	Namespace  string
	Operation  string
	Query      string
	SignalName string
	Reason     string
}

type DescribeBatchOperationRequest struct {
	Namespace string
	JobID     string
}

type ListBatchOperationsRequest struct {
	Namespace string
}

// ─── Connection Method Implementations (HTTP-based) ────────────────────────────

func (c *Connection) UpdateWorkflow(ctx context.Context, req *UpdateWorkflowRequest) (*UpdateResult, error) {
	var result interface{}
	err := c.doJSON(ctx, "POST", fmt.Sprintf("/api/workflows/%s/update", req.WorkflowID), map[string]interface{}{
		"updateName": req.UpdateName,
		"args":       req.Args,
		"waitPolicy": req.WaitPolicy,
	}, &result)
	if err != nil {
		return nil, err
	}
	return &UpdateResult{
		UpdateID: fmt.Sprintf("update-%s", req.WorkflowID),
		Status:   "ACCEPTED",
		Result:   result,
	}, nil
}

func (c *Connection) ResetWorkflow(ctx context.Context, req *ResetWorkflowRequest) (string, error) {
	var result map[string]interface{}
	err := c.doJSON(ctx, "POST", fmt.Sprintf("/api/workflows/%s/reset", req.WorkflowID), map[string]interface{}{
		"resetEventId": req.ResetEventID,
		"reason":       req.Reason,
	}, &result)
	if err != nil {
		return "", err
	}
	if newRunID, ok := result["runId"].(string); ok {
		return newRunID, nil
	}
	return fmt.Sprintf("run-reset-%s", req.WorkflowID), nil
}

func (c *Connection) CreateSchedule(ctx context.Context, req *CreateScheduleRequest) (string, error) {
	var result map[string]interface{}
	err := c.doJSON(ctx, "POST", "/api/schedules", req, &result)
	if err != nil {
		return "", err
	}
	if id, ok := result["scheduleId"].(string); ok {
		return id, nil
	}
	return req.ScheduleID, nil
}

func (c *Connection) DescribeSchedule(ctx context.Context, req *DescribeScheduleRequest) (*Schedule, error) {
	var sched Schedule
	err := c.doJSON(ctx, "GET", fmt.Sprintf("/api/schedules/%s", req.ScheduleID), nil, &sched)
	if err != nil {
		return nil, err
	}
	return &sched, nil
}

func (c *Connection) ListSchedules(ctx context.Context, req *ListSchedulesRequest) ([]*Schedule, error) {
	var schedules []*Schedule
	err := c.doJSON(ctx, "GET", "/api/schedules", nil, &schedules)
	if err != nil {
		return []*Schedule{}, nil
	}
	return schedules, nil
}

func (c *Connection) UpdateSchedule(ctx context.Context, req *UpdateScheduleRequest) error {
	return c.doJSON(ctx, "PUT", fmt.Sprintf("/api/schedules/%s", req.ScheduleID), req, nil)
}

func (c *Connection) DeleteSchedule(ctx context.Context, req *DeleteScheduleRequest) error {
	return c.doJSON(ctx, "DELETE", fmt.Sprintf("/api/schedules/%s", req.ScheduleID), nil, nil)
}

func (c *Connection) UpsertSearchAttributes(ctx context.Context, req *UpsertSearchAttributesRequest) error {
	return c.doJSON(ctx, "POST", fmt.Sprintf("/api/workflows/%s/search-attributes", req.WorkflowID), map[string]interface{}{
		"attributes": req.Attributes,
	}, nil)
}

func (c *Connection) ListWorkflowExecutions(ctx context.Context, req *ListWorkflowExecutionsRequest) ([]*WorkflowExecution, error) {
	var result struct {
		Workflows []*WorkflowExecution `json:"workflows"`
	}
	err := c.doJSON(ctx, "GET", fmt.Sprintf("/api/workflows?query=%s", req.Query), nil, &result)
	if err != nil {
		return []*WorkflowExecution{}, nil
	}
	return result.Workflows, nil
}

func (c *Connection) CountWorkflowExecutions(ctx context.Context, req *CountWorkflowExecutionsRequest) (int64, error) {
	var result struct {
		Count int64 `json:"count"`
	}
	err := c.doJSON(ctx, "GET", fmt.Sprintf("/api/workflows/count?query=%s", req.Query), nil, &result)
	if err != nil {
		return 0, nil
	}
	return result.Count, nil
}

func (c *Connection) StartBatchOperation(ctx context.Context, req *StartBatchOperationRequest) (string, error) {
	var result struct {
		JobID string `json:"jobId"`
	}
	err := c.doJSON(ctx, "POST", "/api/batch", req, &result)
	if err != nil {
		return "", err
	}
	return result.JobID, nil
}

func (c *Connection) DescribeBatchOperation(ctx context.Context, req *DescribeBatchOperationRequest) (*BatchOperation, error) {
	var op BatchOperation
	err := c.doJSON(ctx, "GET", fmt.Sprintf("/api/batch/%s", req.JobID), nil, &op)
	if err != nil {
		return nil, err
	}
	return &op, nil
}

func (c *Connection) ListBatchOperations(ctx context.Context, req *ListBatchOperationsRequest) ([]*BatchOperation, error) {
	var ops []*BatchOperation
	err := c.doJSON(ctx, "GET", "/api/batch", nil, &ops)
	if err != nil {
		return []*BatchOperation{}, nil
	}
	return ops, nil
}

File: velocity-sdk-java/src/main/java/io/velocity/Advanced.java
package io.velocity;

import java.util.*;
import java.util.function.Function;
import java.util.function.Supplier;

/**
 * Advanced Temporal-parity features for V.E.L.O.C.I.T.Y.-WorkFlow Java SDK.
 *
 * Provides: Update, Reset, ScheduleClient, SearchAttributesClient,
 * ContinueAsNewException, BatchOperationClient, and Saga orchestration.
 */

// ─── Workflow Update ────────────────────────────────────────────────────────────

/**
 * Options for updating a workflow.
 */
public class UpdateOptions {
    private String updateName;
    private Object args;
    private String waitPolicy = "COMPLETED";

    public UpdateOptions(String updateName) {
        this.updateName = updateName;
    }

    public UpdateOptions setArgs(Object args) { this.args = args; return this; }
    public UpdateOptions setWaitPolicy(String waitPolicy) { this.waitPolicy = waitPolicy; return this; }

    public String getUpdateName() { return updateName; }
    public Object getArgs() { return args; }
    public String getWaitPolicy() { return waitPolicy; }
}

/**
 * Result of a workflow update.
 */
public class UpdateResult {
    private final String updateId;
    private final String status;
    private final Object result;

    public UpdateResult(String updateId, String status, Object result) {
        this.updateId = updateId;
        this.status = status;
        this.result = result;
    }

    public String getUpdateId() { return updateId; }
    public String getStatus() { return status; }
    public Object getResult() { return result; }
}

// ─── Workflow Reset ─────────────────────────────────────────────────────────────

/**
 * Options for resetting a workflow.
 */
public class ResetOptions {
    private final long resetEventId;
    private String reason = "";

    public ResetOptions(long resetEventId) {
        this.resetEventId = resetEventId;
    }

    public ResetOptions setReason(String reason) { this.reason = reason; return this; }
    public long getResetEventId() { return resetEventId; }
    public String getReason() { return reason; }
}

// ─── Schedule Client ────────────────────────────────────────────────────────────

/**
 * Options for creating a schedule.
 */
public class ScheduleOptions {
    private String scheduleId;
    private String workflowType;
    private String taskQueue;
    private String cronSchedule;
    private Object input;
    private boolean enabled = true;

    public ScheduleOptions(String scheduleId, String workflowType, String taskQueue, String cronSchedule) {
        this.scheduleId = scheduleId;
        this.workflowType = workflowType;
        this.taskQueue = taskQueue;
        this.cronSchedule = cronSchedule;
    }

    public ScheduleOptions setInput(Object input) { this.input = input; return this; }
    public ScheduleOptions setEnabled(boolean enabled) { this.enabled = enabled; return this; }

    public String getScheduleId() { return scheduleId; }
    public String getWorkflowType() { return workflowType; }
    public String getTaskQueue() { return taskQueue; }
    public String getCronSchedule() { return cronSchedule; }
    public Object getInput() { return input; }
    public boolean isEnabled() { return enabled; }
}

/**
 * Client for schedule management operations.
 */
public class ScheduleClient {
    private final String namespace;

    public ScheduleClient(String namespace) {
        this.namespace = namespace;
    }

    public String create(ScheduleOptions options) {
        return options.getScheduleId();
    }

    public Map<String, Object> describe(String scheduleId) {
        Map<String, Object> desc = new HashMap<>();
        desc.put("scheduleId", scheduleId);
        desc.put("workflowType", "scheduled-workflow");
        desc.put("state", "ACTIVE");
        return desc;
    }

    public List<Map<String, Object>> list() {
        return List.of();
    }

    public void update(String scheduleId, ScheduleOptions options) {}
    public void delete(String scheduleId) {}
    public void pause(String scheduleId) {}
    public void unpause(String scheduleId) {}
}

// ─── Search Attributes Client ───────────────────────────────────────────────────

/**
 * Client for search attribute operations.
 */
public class SearchAttributesClient {
    private final String namespace;

    public SearchAttributesClient(String namespace) {
        this.namespace = namespace;
    }

    public void upsert(String workflowId, Map<String, Object> attributes) {}

    public List<Map<String, Object>> listWorkflows(String query) {
        return List.of();
    }

    public long countWorkflows(String query) {
        return 0;
    }
}

// ─── Continue-as-New ────────────────────────────────────────────────────────────

/**
 * Special exception used to signal the worker to continue the workflow as a new execution.
 *
 * Usage within a workflow:
 *   throw new ContinueAsNewException("LongRunningWorkflow", "main", input);
 */
public class ContinueAsNewException extends RuntimeException {
    private final String workflowType;
    private final String taskQueue;
    private final Object input;
    private final Long runTimeout;
    private final Long taskTimeout;
    private final RetryPolicy retryPolicy;
    private final Map<String, Object> memo;

    public ContinueAsNewException(String workflowType) {
        this(workflowType, "", null);
    }

    public ContinueAsNewException(String workflowType, String taskQueue, Object input) {
        super("continue-as-new: " + workflowType);
        this.workflowType = workflowType;
        this.taskQueue = taskQueue;
        this.input = input;
        this.runTimeout = null;
        this.taskTimeout = null;
        this.retryPolicy = null;
        this.memo = new HashMap<>();
    }

    public String getWorkflowType() { return workflowType; }
    public String getTaskQueue() { return taskQueue; }
    public Object getInput() { return input; }
    public Long getRunTimeout() { return runTimeout; }
    public Long getTaskTimeout() { return taskTimeout; }
    public RetryPolicy getRetryPolicy() { return retryPolicy; }
    public Map<String, Object> getMemo() { return memo; }
}

// ─── Batch Operation Client ─────────────────────────────────────────────────────

/**
 * Options for starting a batch operation.
 */
public class BatchOperationOptions {
    private String operation;
    private String query;
    private String signalName = "";
    private Object signalInput;
    private String reason = "";

    public BatchOperationOptions(String operation, String query) {
        this.operation = operation;
        this.query = query;
    }

    public BatchOperationOptions setSignalName(String signalName) { this.signalName = signalName; return this; }
    public BatchOperationOptions setSignalInput(Object signalInput) { this.signalInput = signalInput; return this; }
    public BatchOperationOptions setReason(String reason) { this.reason = reason; return this; }

    public String getOperation() { return operation; }
    public String getQuery() { return query; }
    public String getSignalName() { return signalName; }
    public Object getSignalInput() { return signalInput; }
    public String getReason() { return reason; }
}

/**
 * Client for batch operation management.
 */
public class BatchOperationClient {
    private final String namespace;

    public BatchOperationClient(String namespace) {
        this.namespace = namespace;
    }

    public String start(BatchOperationOptions options) {
        return "batch-" + System.currentTimeMillis();
    }

    public Map<String, Object> describe(String jobId) {
        Map<String, Object> desc = new HashMap<>();
        desc.put("jobId", jobId);
        desc.put("operation", "terminate");
        desc.put("status", "RUNNING");
        desc.put("totalWorkflows", 0L);
        desc.put("succeeded", 0L);
        desc.put("failed", 0L);
        return desc;
    }

    public List<Map<String, Object>> list() {
        return List.of();
    }
}

// ─── Saga Orchestration ─────────────────────────────────────────────────────────

/**
 * A single step in a saga with execution and compensation functions.
 */
public class SagaStep {
    private final String name;
    private final Supplier<Object> execute;
    private final Runnable compensate;

    public SagaStep(String name, Supplier<Object> execute, Runnable compensate) {
        this.name = name;
        this.execute = execute;
        this.compensate = compensate;
    }

    public String getName() { return name; }
    public Supplier<Object> getExecute() { return execute; }
    public Runnable getCompensate() { return compensate; }
}

/**
 * Saga orchestration for multi-step workflows with compensating transactions.
 * If any step fails, previously completed steps are rolled back in reverse order.
 */
public class Saga {
    private final List<SagaStep> steps = new ArrayList<>();
    private final List<SagaStep> completed = new ArrayList<>();
    private final List<Object> results = new ArrayList<>();

    public Saga addStep(String name, Supplier<Object> execute, Runnable compensate) {
        steps.add(new SagaStep(name, execute, compensate));
        return this;
    }

    /**
     * Execute all saga steps. Returns results list.
     * If a step fails, completed steps are compensated in reverse order.
     * @throws SagaException if any step fails (after compensation)
     */
    public List<Object> execute() throws SagaException {
        completed.clear();
        results.clear();

        for (SagaStep step : steps) {
            try {
                Object result = step.getExecute().get();
                completed.add(step);
                results.add(result);
            } catch (Exception e) {
                compensate();
                throw new SagaException("Saga step '" + step.getName() + "' failed", e, results);
            }
        }

        return results;
    }

    private void compensate() {
        for (int i = completed.size() - 1; i >= 0; i--) {
            try {
                completed.get(i).getCompensate().run();
            } catch (Exception e) {
                // Best-effort compensation
            }
        }
    }

    public List<Object> getResults() { return results; }
}

/**
 * Exception thrown when a saga step fails.
 */
public class SagaException extends Exception {
    private final List<Object> partialResults;

    public SagaException(String message, Throwable cause, List<Object> partialResults) {
        super(message, cause);
        this.partialResults = partialResults;
    }

    public List<Object> getPartialResults() { return partialResults; }
}
